File: packages/web/src/algorithms/QC/runQC.ts
import type { Enableable, QCResult, QCRuleResult, QCRulesConfig } from 'src/algorithms/QC/types'
import type { AnalysisResultWithClade, NucleotideSubstitution } from 'src/algorithms/types'

import { getQCRuleStatus } from './QCRuleStatus'
import { ruleMissingData } from './ruleMissingData'
import { ruleMixedSites } from './ruleMixedSites'
import { ruleSnpClusters } from './ruleSnpClusters'
import { rulePrivateMutations } from './rulePrivateMutations'

export type Rule<Conf, Ret> = (
  analysisResult: AnalysisResultWithClade,
  privateMutations: NucleotideSubstitution[],
  config: Conf,
) => Ret

export function runOne<Conf extends Enableable<unknown>, Ret extends QCRuleResult>(
  rule: Rule<Conf, Ret>,
  analysisResult: AnalysisResultWithClade,
  privateMutations: NucleotideSubstitution[],
  config: Conf,
): Ret | undefined {
  return config.enabled ? rule(analysisResult, privateMutations, config) : undefined
}

export interface RunQCParams {
  analysisResult: AnalysisResultWithClade
  privateMutations: NucleotideSubstitution[]
  qcRulesConfig: QCRulesConfig
}

export function runQC({ analysisResult, privateMutations, qcRulesConfig }: RunQCParams): QCResult {
  const result = {
    privateMutations: runOne(rulePrivateMutations, analysisResult, privateMutations, qcRulesConfig.privateMutations),
    missingData: runOne(ruleMissingData, analysisResult, privateMutations, qcRulesConfig.missingData),
    snpClusters: runOne(ruleSnpClusters, analysisResult, privateMutations, qcRulesConfig.snpClusters),
    mixedSites: runOne(ruleMixedSites, analysisResult, privateMutations, qcRulesConfig.mixedSites),
  }

  const score = Object.values(result).reduce((acc, r) => acc + ((r?.score ?? 0) * (r?.score ?? 0)) / 100, 0)

  const status = getQCRuleStatus(score)

  return { seqName: analysisResult.seqName, ...result, overallScore: score, overallStatus: status }
}

File: packages/nextclade/src/qc/runQc.cpp
#include "runQc.h"

#include <nextclade/nextclade.h>

#include <cmath>
#include <functional>

#include "getQcRuleStatus.h"

namespace Nextclade {

  template<typename T>
  double addScore(const std::optional<T>& ruleResult) {
    if (ruleResult) {
      return std::pow(ruleResult->score, 2) * 0.01;
    }
    return 0.0;
  }

  QcResult runQc(                              //
    const ::NextalignResultInternal& alignment,//
    const AnalysisResult& analysisResult,      //
    const QcConfig& qcRulesConfig              //
  ) {
    QcResult result = {
      .missingData = ruleMissingData(analysisResult, qcRulesConfig.missingData),
      .mixedSites = ruleMixedSites(analysisResult, qcRulesConfig.mixedSites),
      .privateMutations = rulePrivateMutations(analysisResult, qcRulesConfig.privateMutations),
      .snpClusters = ruleSnpClusters(analysisResult, qcRulesConfig.snpClusters),
      .frameShifts = ruleFrameShifts(analysisResult, qcRulesConfig.frameShifts),
      .stopCodons = ruleStopCodons(alignment, qcRulesConfig.stopCodons),
      .overallScore = 0,              // Will be overwritten below
      .overallStatus = QcStatus::good,// Will be overwritten below
    };

    result.overallScore += addScore(result.missingData);
    result.overallScore += addScore(result.mixedSites);
    result.overallScore += addScore(result.privateMutations);
    result.overallScore += addScore(result.snpClusters);
    result.overallScore += addScore(result.frameShifts);
    result.overallScore += addScore(result.stopCodons);

    result.overallStatus = getQcRuleStatus(result.overallScore);

    return result;
  }
}// namespace Nextclade
